doc/dev/ffi.md

```diff
@@ -53,10 +53,59 @@ In the case of `@[extern]` all *irrelevant* types are removed first; see next se
   Its runtime value is either a pointer to an opaque bignum object or, if the lowest bit of the "pointer" is 1 (`lean_is_scalar`), an encoded unboxed natural number (`lean_box`/`lean_unbox`).
 * A universe `Sort u`, type constructor `... → Sort u`, or proposition `p : Prop` is *irrelevant* and is either statically erased (see above) or represented as a `lean_object *` with the runtime value `lean_box(0)`
 * Any other type is represented by `lean_object *`.
-  Its runtime value is a pointer to an object of a subtype of `lean_object` (see respective declarations in `lean.h`) or the unboxed value `lean_box(cidx)` for the `cidx`th constructor of an inductive type if this constructor does not have any relevant parameters.
+  Its runtime value is a pointer to an object of a subtype of `lean_object` (see the "Inductive types" section below) or the unboxed value `lean_box(cidx)` for the `cidx`th constructor of an inductive type if this constructor does not have any relevant parameters.
 
   Example: the runtime value of `u : Unit` is always `lean_box(0)`.
 
+#### Inductive types
+
+For inductive types which are in the fallback `lean_object *` case above and not trivial constructors, the type is stored as a `lean_ctor_object`, and `lean_is_ctor` will return true. A `lean_ctor_object` stores the constructor index in the header, and the fields are stored in the `m_objs` portion of the object.
+
+The memory order of the fields is derived from the types and order of the fields in the declaration. They are ordered as follows:
+
+* Non-scalar fields stored as `lean_object *`
+* Fields of type `USize`
+* Other scalar fields, in decreasing order by size
+
+Within each group the fields are ordered in declaration order. **Warning**: Trivial wrapper types still count toward a field being treated as non-scalar for this purpose.
```

Comment:
I did not know this until I sat down to document it. I'm pretty sure this is a bug.

Reply:
So what happens with the stored value, is it boxed or extended (what happens with a trivial UInt64 wrapper on 32 bit?)?

Reply:
It's boxed. Basically it's the same as the fallback encoding for e.g. the elements of `Array UInt32`, which stores its values in `lean_object*` slots and uses `lean_box` and `lean_unbox` to get at the values.

We should probably say something about how types like `Array UInt32` work here...

Reply:
As for UInt64 on 32 bit, I'm actually not sure, I was also wondering about that. The current scheme of putting `usize, u64, u32, u16, u8` in that order and not using any padding seems to assume that `usize` has the largest alignment, but I don't think this is true on 32 bit unless u64 is also 32 bit aligned, which I think is not the case on at least some of the supported arches. From what I can tell of the code though there is no guard against this condition, the offsets are still as documented here and the u64s will be unaligned in this case.